Add new function templates dropdown (#3486)
Qiskit/documentation: docs/guides/qiskit-function-templates.mdx

---
title: Qiskit Function templates
description: Learn how to use and build Qiskit Function templates.
---

# Introduction to Qiskit Function templates

> [Qiskit Function templates](https://github.com/qiskit-community/qiskit-function-templates) are an open-source collection of realistic code examples that give you access to some of the latest research workflows that leverage tools like the Qiskit SDK and [Qiskit addons](https://quantum.cloud.ibm.com/docs/en/guides/addons). These pre-built workflows take learnings from practical experiments and package them in a way that you can customize and reuse in your own research so that you don't have to build from scratch. All templates are readily deployable to [Qiskit Serverless](./serverless) where you can easily leverage classical compute as part of running your workflow.

Reviewer: ```suggestion
[Qiskit Function templates](https://github.com/qiskit-community/qiskit-function-templates) are an open-source collection of realistic code examples that give you access to some of the latest research workflows that leverage tools like the Qiskit SDK and [Qiskit addons](https://quantum.cloud.ibm.com/docs/en/guides/addons). These pre-built workflows take information from practical experiments and package them in a way that you can customize and reuse in your own research so that you don't have to build from scratch. All templates are readily deployable to [Qiskit Serverless](./serverless) where you can easily leverage classical compute as part of your workflow.
```

> Templates are available in two varieties:

Reviewer: ```suggestion
There are two types of templates:
```

- *Template implementations*: Specialized code examples that implement advanced techniques from real-world experiments in areas like chemistry and physics.

> - *Base templates*: References for developing your own original Qiskit Function that highlight best-practices in interface development, code formatting, unit testing, and more.

Reviewer: ```suggestion
- *Base templates*: References for developing your own original Qiskit Function. These templates highlight best-practices in interface development, code formatting, unit testing, and more.
```

## Template implementations

Qiskit Function template implementations are organized by application area. Currently included in the collection is a physics template for Hamiltonian simulation using the [AQC-Tensor Qiskit addon](./qiskit-addons-aqc) and a chemistry template for electronic structure with the implicit solvent model using the [SQD Qiskit addon](./qiskit-addons-sqd).

> Resources to get started with these two templates are available at the links below:

Reviewer: ```suggestion
Resources to get started with these two templates are available at the following links:
```

- Electronic structure simulation with implicit solvent model: [template source files](https://github.com/qiskit-community/qiskit-function-templates/tree/main/chemistry/sqd_pcm) and [guide](./function-template-chemistry-workflow)
- Hamiltonian simulation: [template source files](https://github.com/qiskit-community/qiskit-function-templates/tree/main/physics/hamiltonian_simulation) and [guide](./function-template-hamiltonian-simulation)

## Base templates

> The *base templates* are references for developing your own original Qiskit Function. They are designed to help you kickstart the development process by illustrating best practices in interface development, code formatting, I/O handling, error messages, checkpoints, unit testing, and more. The repository itself is exposed as a [GitHub template](https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-template-repository) so that the CI/CD implementation can be easily extended.

Reviewer: ```suggestion
The *base templates* are references for developing your own original Qiskit Function. They are designed to help you kickstart the development process by illustrating best practices in interface development, code formatting, I/O handling, error messages, checkpoints, unit testing, and more. The repository is exposed as a [GitHub template](https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-template-repository) so that the CI/CD implementation can be easily extended.
```

> There are currently two templates, a [circuit function template](https://github.com/qiskit-community/qiskit-function-templates/blob/main/base_templates/circuit_function_template.py) and an [application function template](https://github.com/qiskit-community/qiskit-function-templates/blob/main/base_templates/application_function_template.py). The circuit function template is built similarly to the [IBM Circuit Function](./ibm-circuit-function) and the application function template illustrates how to structure a typical application workflow.

Reviewer: ```suggestion
There are currently two templates: a [circuit function template](https://github.com/qiskit-community/qiskit-function-templates/blob/main/base_templates/circuit_function_template.py) and an [application function template](https://github.com/qiskit-community/qiskit-function-templates/blob/main/base_templates/application_function_template.py). The circuit function template is built similarly to the [IBM Circuit Function](./ibm-circuit-function) and the application function template illustrates how to structure a typical application workflow.
```

## Base templates

The *base templates* are references for developing your own original Qiskit Function. They are designed to help you kickstart the development process by illustrating best practices in interface development, code formatting, I/O handling, error messages, checkpoints, unit testing, and more. The repository itself is exposed as a [GitHub template](https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-template-repository) so that the CI/CD implementation can be easily extended.

There are currently two templates, a [circuit function template](https://github.com/qiskit-community/qiskit-function-templates/blob/main/base_templates/circuit_function_template.py) and an [application function template](https://github.com/qiskit-community/qiskit-function-templates/blob/main/base_templates/application_function_template.py). The circuit function template is built similarly to the [IBM Circuit Function](./ibm-circuit-function) and the application function template illustrates how to structure a typical application workflow.

## Next steps

<Admonition type="info" title="Recommendations">

> - Review the guide on building a function template for [Hamiltonian simulation](./function-template-hamiltonian-simulation)

Reviewer 2: Add link to the Chemistry guide page as well as the Ham sim guide page

---
title: Qiskit Function templates
description: Learn how to use Qiskit Code Assistant, a generative AI code assistant.
---
{/* cspell:ignore FCIDUMP */}

# Introduction to Qiskit Function templates

The [Qiskit Function templates](https://github.com/qiskit-community/qiskit-function-templates) provide realistic examples for how to build utility-scale application workflows with Qiskit and the Qiskit addons. These prebuilt workflows are customizable for a variety of research needs and can be used instead of starting from scratch. To simplify execution and parallelization, they are easily deployable to [Qiskit Serverless](./serverless).

> The templates provided are broken into two basic categories:

Reviewer 3: Maybe it's worth walking through what's included in a Base Template pipeline, so you can further illustrate what's interesting about the Application template?

- *Application Templates*: Templates to build realistic research workflows.
- *Base Templates*: Basic implementation use for reference in the development process.

## Application template implementations

The application template implementations are organized by application area, which currently includes a physics and chemistry application utilizing the [SQD addon](./qiskit-addons-sqd) and the integral equation formalism polarizable continuum model (IEF-PCM).

For convenience, direct download links are provided for each application template directory:

- [Download Hamiltonian Simulation template](https://ibm.biz/ham-sim-template)
- [Download SQD IEF-PCM template](https://ibm.biz/sqd-pcm-template)

### Chemistry Workflow

> This template encapsulates a workflow for quantum-centric implicit solvent simulations of ground state energy in molecular systems. More detail about the workflow's methodology can be found in the template's [README.md documentation](https://github.com/qiskit-community/qiskit-function-templates/blob/main/chemistry/sqd_pcm/README.md) and a Jupyter notebook describing how to deploy and run this template can be found in the [same directory](https://github.com/qiskit-community/qiskit-function-templates/blob/main/chemistry/sqd_pcm/deploy_and_run.ipynb). It is composed of three sections described below:

Reviewer 3: I think this section have a few details on interesting things users should experiment with, in line with "customisable pipelines", where you can replace components yourself (rather than reconfigurations)

2. Mapping

The input arguments are then used to construct the relevant quantum circuits and operators for the workflow. A target circuit is created and a matrix product state representation of this circuit is found using the AQC addon. Following this, an ansatz circuit is generated and optimized using tensor network methods, producing a final circuit which executes the remainder of the time evolution.

3. Execution

Finally, the ansatz circuit is transpiled and executed on a QPU and collects estimates for all of the specified expectation values, which are returned in a serializable format for access by the user.

## Base templates

The *base templates* are reference implementations to kickstart the Qiskit Function development process with best-practices in interface development, code formatting, unit testing, and more. The repository itself is exposed as a [GitHub template](https://docs.github.com/en/repositories/creating-and-managing-repositories/creating-a-template-repository) so that the CI/CD implementation can be easily extended.

> There are currently two templates, a [circuit function template](https://github.com/qiskit-community/qiskit-function-templates/blob/main/base_templates/circuit_function_template.py) and an [application function template](https://github.com/qiskit-community/qiskit-function-templates/blob/main/base_templates/application_function_template.py). The circuit function template is built similarly to the [IBM Circuit Function](./ibm-circuit-function) and the application function template provides a structure to build an application workflow.

Reviewer 2: ```suggestion
There are currently two templates, a [circuit function template](https://github.com/qiskit-community/qiskit-function-templates/blob/main/base_templates/circuit_function_template.py) and an [application function template](https://github.com/qiskit-community/qiskit-function-templates/blob/main/base_templates/application_function_template.py). The circuit function template is built similarly to the [IBM Circuit Function](./ibm-circuit-function) and the application function template illustrates how to structure a typical application workflow.
```
